### PR TITLE
New hardware file format

### PR DESCRIPTION
**Pull request by egthomas**
This pull request adds the new-format `hdw.dat` files proposed by the Tdiff Task Force to the SuperDARN PIs and makes the necessary C and IDL software changes needed to use them.  Changes include

- human-readable dates and times for each line of the hdw files
- a new status code in the hdw files to indicate the radar operations status (ie allows for radars to come and go, like HAL or FIR)
- a new electronic beam offset field in the hdw files to account for radars which have electronically shifted the scanning boresight away from the physical antenna boresight (eg WAL, BKS), which is important for the elevation angle determination
- a separate `tdiff` value for each channel of a Stereo radar (eg HAN, PYK); single-channel radars will only use the first `tdiff` value

Necessary changes have been made to the FITACF, elevation angle, and geolocation software to use these new multi-channel `tdiff` values and electronic beam offsets.  No other information in the hardware files should have been changed from the previous set of files.

Note that we are opening this as a draft pull request first so that users have a chance to test the new software and verify the contents of the new hardware files.  An equivalent pull request will be made to the official hdw repository once any issues are addressed here.

### CODE REVIEW

**Comment by egthomas** [comment]
> Right now I am looking through the README document and having a feeling that it might need a bit more wordsmithing in order to make the content more precise/clear/comprehendable.

> I can make suggestions in a marked-up Word document and post it here for further discusison. Would this be OK or a separate issue would be more appropriate?

@pasha-ponomarenko any help with revisions to the README file would be greatly appreciated, either here or tracked in a separate issue.  My personal feeling is that I don't think we should let README edits slow down this pull request, as the main purpose is to test the software compatibility with the new hardware files.

File: tables/superdarn/hdw/README.txt [hunk 1]
```diff
@@ -15,36 +15,28 @@ delineated by one or more spaces. The following table specifies the
...
-05) Geographic longitude of radar site (Given in decimal degrees to
+06) Geographic longitude of radar site (Given in decimal degrees to
     3 decimal places. West longitude values are negative)
-06) Altitude of the radar site (meters)
-07) Scanning boresight (Direction of the center beam, measured in
+07) Altitude of the radar site (meters)
+08) Physical scanning boresight (Direction of the center beam, measured in
     degrees relative to geographic north. CCW rotations are negative.)
-08) Beam separation (Angular separation in degrees between adjacent
+09) Electronic shift to radar scanning boresight (Degrees relative to
```
Comment (ksterne) [inline tables/superdarn/hdw/README.txt:28]:
```suggestion
09) Shift to radar physical scanning boresight (Degrees relative to
```
Suggestion to remove electronic to reduce confusion.

File: tables/superdarn/hdw/hdw.dat.sye [hunk 2]
```diff
@@ -6,8 +6,9 @@
...
 #
-#  1   2      3        4         5       6      7      8    9   10   11   12  13(1)  13(2)  13(3)  14   15  16 17
-#
-# UNTIL 01/Jan/3000 05:45:36.0
-   13 2999 31556736  -69.010   39.610    42.5  106.5 -3.33  1   10  0.000  1   -5.0   92.0    0.8  50.0  3  75 16
+#                            GEOGRAPHIC             BORE-   OFF  BEAM V  |--------- interferometer ---------| |--- RX ---| RNG MX
+#      YYYYMMDD HH:MM:SS    LAT      LON      ALT   SIGHT   SET  SEP  SG PS  TDIFF  TDIFF    X      Y     Z    RXRIS  ATTN     BM
+ 13  1 19970208 00:00:00  -69.010   39.610    42.5  106.5  0.00 -3.33  1  1  0.000  0.000   -5.0   92.0   0.8   50.0  10 3  75 16
+ 13  1 19971227 08:17:00  -69.010   39.610    42.5  106.5  0.00 -3.33 -1  1  0.000  0.000   -5.0   92.0   0.8   50.0  10 3  75 16
+ 13  1 19990716 11:18:29  -69.010   39.610    42.5  106.5  0.00 -3.33  1  1  0.000  0.000   -5.0   92.0   0.8   50.0  10 3  75 16
```
Comment (aburrell) [inline tables/superdarn/hdw/hdw.dat.sye:13]: Confirming this is an update to the hardware file.

File: tables/superdarn/hdw/hdw.dat.sys [hunk 3]
```diff
@@ -6,11 +6,10 @@
...
-   12 1999        0  -69.000   39.580     0.0  159.0 -3.33  1   10  0.000  1   -5.0   92.0    0.8  50.0  3  75 16
-#
-# UNTIL 01/Jan/3000 05:45:36.0
-   12 2999 31556736  -69.000   39.580     0.0  165.0 -3.33  1   10  0.000  1   -5.0   92.0    0.8  50.0  3  75 16
+#                            GEOGRAPHIC             BORE-   OFF  BEAM V  |--------- interferometer ---------| |--- RX ---| RNG MX
+#      YYYYMMDD HH:MM:SS    LAT      LON      ALT   SIGHT   SET  SEP  SG PS  TDIFF  TDIFF    X      Y     Z    RXRIS  ATTN     BM
+ 12  1 19950201 00:00:00  -69.000   39.580     0.0  159.0  0.00 -3.33 -1  1  0.000  0.000   -5.0   92.0   0.8   50.0  10 3  75 16
+ 12  1 19960424 11:00:00  -69.000   39.580     0.0  159.0  0.00 -3.33  1  1  0.000  0.000   -5.0   92.0   0.8   50.0  10 3  75 16
+ 12  1 19990101 00:00:00  -69.000   39.580     0.0  159.0  6.00 -3.33  1  1  0.000  0.000   -5.0   92.0   0.8   50.0  10 3  75 16
+ 12  1 20001231 06:00:00  -69.000   39.580     0.0  159.0  0.00 -3.33  1  1  0.000  0.000   -5.0   92.0   0.8   50.0  10 3  75 16
```
Comment (aburrell) [inline tables/superdarn/hdw/hdw.dat.sys:14]: Confirming the updates are intended.

**Review by aburrell — COMMENTED** [review]
There are some updates to the hardware files that I want to confirm were intended.  Overall, it looks good to me.  I will try to merge this into the make_fov branch and get it to work.

File: tables/superdarn/hdw/hdw.dat.sps [hunk 4]
```diff
@@ -6,10 +6,8 @@
...
 #
-#  1   2      3        4         5       6      7      8    9   10   11   12  13(1)  13(2)  13(3)  14   15  16 17
-#
-# UNTIL 01/Jan/2020 02:46:40.0
-   22 2020 10000  -89.995  118.291  3000.0   75.7  3.24   1   10  0.025 -1    0.0  -97.5    0.0   0.0  2  75 16
-# UNTIL 01/Jan/2999 02:46:40.0
-   22 2999 10000  -89.995  118.291  3000.0   75.7  3.24  -1   10  0.025 -1    0.0  -97.5    0.0   0.0  2  75 16
+#                            GEOGRAPHIC             BORE-   OFF  BEAM V  |--------- interferometer ---------| |--- RX ---| RNG MX
+#      YYYYMMDD HH:MM:SS    LAT      LON      ALT   SIGHT   SET  SEP  SG PS  TDIFF  TDIFF    X      Y     Z    RXRIS  ATTN     BM
+ 22  1 20130122 00:00:00  -89.995  118.291  3000.0   75.7  0.00  3.24  1  1  0.025  0.000    0.0  -97.5   0.0    0.0  10 2  75 16
```
[the remark this answers is not in the document]
Comment (egthomas) [reply tables/superdarn/hdw/hdw.dat.sps:11]: You're right, somewhere along the way this was accidentally changed; both phase sign values for SPS should be -1.  I'll push a fix soon.

File: tables/superdarn/hdw/hdw.dat.sys (hunk 3 above)
Comment (egthomas) [reply tables/superdarn/hdw/hdw.dat.sys:14]: Good catch @aburrell - I forgot that these changes were implemented.  I reached out to Sessai earlier in the summer and am still waiting on a response, but based on the information from these pages I feel more confident in the revised versions on this branch than the current hardware files for Syowa South and East:

http://polaris.nipr.ac.jp/~SD/sensu/SW/IQreverse/
https://web.archive.org/web/20070719171647/http://www.uap.nipr.ac.jp/SD/SyowaSouthNew/

**Comment by egthomas** [comment]
Ok, it's been 4 weeks since first making these changes available as a draft pull request so I'm going to mark this as "ready for review".  Comments or feedback are welcome.

**Comment by ecbland** [comment]
@egthomas I know this PR has been sitting here for a while -- do you want any further PI input (apart from Sessai approving the Syowa changes)? I can request this in the next DAWG report (due on November 15). We probably need to be specific about what kind of input is needed.

**Comment by pasha-ponomarenko** [comment]
Back in August I have some comments and suggested changes to the readme content but failed to share them. Please see those attached. In general, there are several mute points like, e.g., array normal/center beam/boresight definitions, which need to be clarified. Also, having comments embedded into field descriptions looks a bit out of place, and there are some duplicates like "20) Analog attenuation **stages** (Number of **stages**.
[Format readme comments from Pasha.docx](https://github.com/SuperDARN/rst/files/7514730/Format.readme.comments.from.Pasha.docx)
 "

**Comment by aburrell** [comment]
Recommend capturing the documentation changes in an issue, so the hardware changes can be merged and a full dicussion devoted to the documentation.

**Comment by pasha-ponomarenko** [comment]
Thanks @aburrell, I will create a separate issue.
:-)

**Comment by egthomas** [comment]
I've just received permission from the King Salmon (KSR) radar PI (Bill Bristow) to correct an error in the interferometer Y-offset value (from +100 m to -100 m), so I'll make that change here (rather than waiting on #466 which has had some valid concerns raised).

**Comment by egthomas** [comment]
We've now received permission from the SANAE PI (Mike Kosch) to correct a similar error in that radar's interferometer Y-offset value (again from +100 to -100 m), so I'll make that correction on this branch as well.

**Comment by egthomas** [comment]
We've now received confirmation from the Syowa East / South PI (Sessai) that the velocity sign and electronic offset changes we made are appropriate.  Note there may still be some updates necessary to the SYS interferometer X- and Z-offset values.

**Comment by ecbland** [comment]
@egthomas Could you please add a modification line to each `.c`, `.h` and `.pro` file you have modified in this PR?

File: codebase/superdarn/src.dlm/rposdlm.1.9/rposdlm.c [hunk 5]
```diff
@@ -511,20 +528,23 @@ static IDL_VPTR IDLRadarYMDHMSGetSite(int argc,IDL_VPTR *argv) {
 
   isite=IDLRadarMakeSite(&vsite);
 
+  isite->tval=status->site[s].status;
```
Comment (ecbland) [inline codebase/superdarn/src.dlm/rposdlm.1.9/rposdlm.c:531]: This line is causing a compilation error (see below). It compiles ok with the following change: 

```suggestion
  isite->tval=iradar->site[s].status;
```

```
================================================================================
Compiling Binary:/rst/codebase/superdarn/src.dlm/rposdlm.1.9
================================================================================
make clean
rm -f *.o
make
make.version /home/emma/gitrepos/rst/codebase/superdarn/src.dlm/rposdlm.1.9/..
cc -fPIC -Wall -pedantic -O3 -D_GNU_SOURCE -D_LINUX -I/rst/include/base -I/rst/include/general -I/rst/include/superdarn -I/usr/local/itt/idl/external/include -c rposdlm.c    
rposdlm.c: In function ‘IDLRadarYMDHMSGetSite’:
rposdlm.c:531:15: error: ‘status’ undeclared (first use in this function); did you mean ‘statx’?
  531 |   isite->tval=status->site[s].status;
      |               ^~~~~~
      |               statx
rposdlm.c:531:15: note: each undeclared identifier is reported only once for each function it appears in
make: *** [/rst/build/make/makedlm.linux:19: rposdlm.o] Error 1
Compilation Aborted.
```

Comment (egthomas) [reply codebase/superdarn/src.dlm/rposdlm.1.9/rposdlm.c:531]: @ecbland thanks for testing!

**Comment by egthomas** [comment]
> @egthomas Could you please add a modification line to each .c, .h and .pro file you have modified in this PR?

@ecbland those lines should now be included in all of the modified .c, .h., and .pro files - let me know if I missed any.

**Comment by pasha-ponomarenko** [comment]
@egthomas , @ecbland , I was testing utilisation of the two _tdiff_ values for calibrating non-stereo two-channel data and came across an apparent problem.  It seems that there is a mix-up with applying _tdiff_ values to stereo data by FITACF2.5. If I change _tdiff_[0], elevation values  change in both channels but if I change _tdiff_[1], there are no changes in either of them. This problem does not exist for the data processed with FITACF3 when changes in _tdiff_[0] cause changes in channel 1 only, and changes in _tdiff_[1] lead to changes in channel 2 only. Please see attached a stack of PPT slides illustrating this issue. 

[TDIFF misfit for FITACF2.pptx](https://github.com/SuperDARN/rst/files/8347729/TDIFF.misfit.for.FITACF2.pptx)

I used _develop_ branch for this analysis.

I suspect that this issue may be caused by the way tdiff is declared in FitPrm structure (a single value instead of a two-element array), but I can easily be mistaken.

**Pull request merged**